File: reference-implementations/dotnet/TomlSchema/SchemaDiscovery.cs
namespace TomlSchema;

using Tomlyn;
using Tomlyn.Model;
using System.Text.RegularExpressions;

/// <summary>
/// Discovers a schema referenced by a TOML document's reserved
/// <c>[toml-schema].location</c>, following the resolution and
/// version-compatibility rules of SPEC.md's
/// "TOML Reference of a TOML Schema" section.
/// </summary>
internal static class SchemaDiscovery
{
    private static readonly Regex SemVerPattern = new(
        @"^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)"
        + @"(?:-((?:0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*)"
        + @"(?:\.(?:0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*))*))?"
        + @"(?:\+([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?$",
        RegexOptions.Compiled
    );

    public static DiscoveredSchema Discover(string documentPath)
    {
        var content = File.ReadAllText(documentPath);
        var document = TomlSerializer.Deserialize<TomlTable>(content)
            ?? throw new InvalidOperationException($"Failed to parse document: {documentPath}");

        if (!document.TryGetValue("toml-schema", out var metadataObj) || metadataObj is not TomlTable metadata)
            throw new InvalidOperationException("document does not contain [toml-schema].location");

        foreach (var (key, value) in metadata)
        {
            if (!IsScalar(value))
                throw new InvalidOperationException($"document [toml-schema].{key} must be a scalar value");
        }

        if (!metadata.TryGetValue("location", out var locationObj) || locationObj is not string location
            || string.IsNullOrWhiteSpace(location))
            throw new InvalidOperationException("document does not contain [toml-schema].location");

        var schemaPath = ResolveSchemaLocation(documentPath, location.Trim());
        var schema = TomlSchema.Load(schemaPath);

        var warnings = new List<ValidationDiagnostic>();
        if (metadata.TryGetValue("version", out var versionObj))
        {
            if (versionObj is not string expectedVersion)
                throw new InvalidOperationException("document [toml-schema].version must be a SemVer string");

            var expectedMatch = SemVerPattern.Match(expectedVersion);
            if (!expectedMatch.Success)
                throw new InvalidOperationException(
                    "document [toml-schema].version must use SemVer MAJOR.MINOR.PATCH syntax");

            var actualMatch = SemVerPattern.Match(schema.Version);
            var expectedMajor = expectedMatch.Groups[1].Value;
            var actualMajor = actualMatch.Groups[1].Value;

            if (expectedMajor != actualMajor)
            {
                throw new InvalidOperationException(
                    $"document expects TOML Schema major version {expectedVersion}, "
                    + $"but resolved schema uses {schema.Version}");
            }

            if (expectedVersion != schema.Version)
            {
                warnings.Add(new ValidationDiagnostic(
                    DiagnosticPhase.Discovery,
                    DiagnosticSeverity.Warning,
                    DiagnosticCodes.VersionMismatch,
                    null,
                    null,
                    $"document expects TOML Schema version {expectedVersion}, but resolved schema uses {schema.Version}"));
            }
        }

        return new DiscoveredSchema(schema, document, warnings);
    }

    private static bool IsScalar(object? value) =>
        value is not TomlTable && value is not TomlArray && value is not TomlTableArray;

    private static string ResolveSchemaLocation(string documentPath, string location)
    {
        if (IsAbsoluteLocalPath(location))
            return Path.GetFullPath(location);

        if (HasInvalidUriReferenceCharacter(location))
            throw new InvalidOperationException($"invalid [toml-schema].location URI: {location}");

        Uri reference;
        try
        {
            reference = new Uri(location, UriKind.RelativeOrAbsolute);
        }
        catch (UriFormatException ex)
        {
            throw new InvalidOperationException($"invalid [toml-schema].location URI: {location}: {ex.Message}");
        }

        var absoluteDocumentPath = Path.GetFullPath(documentPath);
        var baseUri = new Uri("file://" + ToUriPath(absoluteDocumentPath), UriKind.Absolute);
        Uri resolved;
        try
        {
            resolved = reference.IsAbsoluteUri ? reference : new Uri(baseUri, reference);
        }
        catch (UriFormatException)
        {
            // A malformed absolute reference, such as an opaque "file:schema.tosd" URI that lacks
            // the authority component a hierarchical file URI requires, fails during resolution.
            throw new InvalidOperationException($"invalid file schema location: {location}");
        }

        if (!string.Equals(resolved.Scheme, "file", StringComparison.OrdinalIgnoreCase))
            throw new InvalidOperationException($"unsupported schema location URI scheme: {resolved.Scheme}");

        return LocalPathFromFileUri(location, resolved);
    }

    private static string ToUriPath(string absolutePath)
    {
        var normalized = absolutePath.Replace('\\', '/');
        return normalized.StartsWith('/') ? normalized : "/" + normalized;
    }

    private static bool IsAbsoluteLocalPath(string location)
    {
        if (location.StartsWith('/'))
            return true;
        if (location.Length >= 3 && char.IsLetter(location[0]) && location[1] == ':'
            && (location[2] == '\\' || location[2] == '/'))
            return true;
        return false;
    }

    private static bool HasInvalidUriReferenceCharacter(string reference)
    {
        foreach (var character in reference)
        {
            if (character <= ' ' || character == 0x7f)
                return true;
            switch (character)
            {
                case '\\':
                case '"':
                case '<':
                case '>':
                case '^':
                case '`':
                case '{':
                case '|':
                case '}':
                    return true;
            }
        }
        return false;
    }

    private static string LocalPathFromFileUri(string location, Uri uri)
    {
        if (!string.IsNullOrEmpty(uri.UserInfo) || !string.IsNullOrEmpty(uri.Query)
            || !string.IsNullOrEmpty(uri.Fragment))
            throw new InvalidOperationException($"invalid file schema location: {location}");

        var host = uri.Host;
        if (!string.IsNullOrEmpty(host) && !string.Equals(host, "localhost", StringComparison.OrdinalIgnoreCase))
            throw new InvalidOperationException($"file URI has a non-local host: {location}");

        // System.Uri decodes %2F/%5C while parsing, so an encoded separator can only be detected
        // by inspecting the original, still-escaped [toml-schema].location text.
        var lowerLocation = location.ToLowerInvariant();
        if (lowerLocation.Contains("%2f") || lowerLocation.Contains("%5c"))
            throw new InvalidOperationException($"file URI contains an encoded path separator: {location}");

        // AbsolutePath (not LocalPath) is used because LocalPath renders a UNC-style
        // "\\host\path" form whenever a host component is present, even a benign "localhost".
        var path = uri.AbsolutePath;
        if (string.IsNullOrEmpty(path) || path.Contains('\0'))
            throw new InvalidOperationException($"file URI does not contain a safe path: {location}");

        if (!Path.IsPathRooted(path))
            throw new InvalidOperationException($"file URI path is not absolute: {location}");

        return Path.GetFullPath(path);
    }
}

File: reference-implementations/java/src/main/java/org/tomlschema/SchemaDiscovery.java
package org.tomlschema;

import org.tomlj.Toml;
import org.tomlj.TomlArray;
import org.tomlj.TomlParseResult;
import org.tomlj.TomlTable;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Discovers a schema referenced by a TOML document's reserved
 * {@code [toml-schema].location}, following the resolution and
 * version-compatibility rules of SPEC.md's
 * "TOML Reference of a TOML Schema" section.
 */
final class SchemaDiscovery {
    private SchemaDiscovery() {
    }

    static DiscoveredSchema discover(Path documentPath) throws IOException {
        TomlParseResult document = Toml.parse(documentPath);
        if (document.hasErrors()) {
            throw new SchemaException("Unable to parse document " + documentPath + ": "
                    + formatParseErrors(document));
        }

        TomlTable metadata = document.getTable("toml-schema");
        if (metadata == null) {
            throw new SchemaException("document does not contain [toml-schema].location");
        }
        for (String key : metadata.keySet()) {
            if (!isScalar(metadata.get(key))) {
                throw new SchemaException("document [toml-schema]." + key + " must be a scalar value");
            }
        }
        Object rawLocation = metadata.get("location");
        String location = rawLocation instanceof String value ? value.strip() : "";
        if (location.isEmpty()) {
            throw new SchemaException("document does not contain [toml-schema].location");
        }

        Path schemaPath = resolveSchemaLocation(documentPath, location);
        TomlSchema schema = TomlSchema.load(schemaPath);

        List<ValidationDiagnostic> warnings = new ArrayList<>();
        if (metadata.contains("version")) {
            TomlSchemaVersion.Version expected = TomlSchemaVersion.parseDocumentVersion(metadata.get("version"));
            TomlSchemaVersion.Version actual = TomlSchemaVersion.parseDocumentVersion(schema.version());
            if (!expected.major().equals(actual.major())) {
                throw new SchemaException(DiagnosticPhase.DISCOVERY, DiagnosticCodes.UNSUPPORTED_VERSION,
                        "$.toml-schema.version",
                        "document expects TOML Schema major version " + expected.value()
                        + ", but resolved schema uses " + schema.version());
            }
            if (!expected.value().equals(schema.version())) {
                warnings.add(ValidationDiagnostic.warning(DiagnosticPhase.DISCOVERY,
                        DiagnosticCodes.VERSION_MISMATCH, null, "$.toml-schema.version",
                        "document expects TOML Schema version " + expected.value()
                                + ", but resolved schema uses " + schema.version()));
            }
        }

        return new DiscoveredSchema(schema, document, warnings);
    }

    private static boolean isScalar(Object value) {
        return !(value instanceof TomlArray) && !(value instanceof TomlTable);
    }

    private static Path resolveSchemaLocation(Path documentPath, String location) {
        if (isAbsoluteLocalPath(location)) {
            return Path.of(location).normalize();
        }
        if (hasInvalidUriReferenceCharacter(location)) {
            throw new SchemaException("invalid [toml-schema].location URI: " + location);
        }
        URI reference;
        try {
            reference = new URI(location);
        } catch (URISyntaxException e) {
            throw new SchemaException("invalid [toml-schema].location URI: " + location + ": " + e.getMessage());
        }
        URI base = documentPath.toAbsolutePath().normalize().toUri();
        URI resolved = base.resolve(reference);
        if (!"file".equalsIgnoreCase(resolved.getScheme())) {
            throw new SchemaException("unsupported schema location URI scheme: " + resolved.getScheme());
        }
        return localPathFromFileUri(location, resolved);
    }

    private static boolean isAbsoluteLocalPath(String location) {
        if (location.startsWith("/")) {
            return true;
        }
        if (location.length() >= 3 && Character.isLetter(location.charAt(0)) && location.charAt(1) == ':'
                && (location.charAt(2) == '\\' || location.charAt(2) == '/')) {
            return true;
        }
        return false;
    }

    private static boolean hasInvalidUriReferenceCharacter(String reference) {
        for (int i = 0; i < reference.length(); i++) {
            char character = reference.charAt(i);
            if (character <= ' ' || character == 0x7f) {
                return true;
            }
            switch (character) {
                case '\\':
                case '"':
                case '<':
                case '>':
                case '^':
                case '`':
                case '{':
                case '|':
                case '}':
                    return true;
                default:
                    break;
            }
        }
        return false;
    }

    private static Path localPathFromFileUri(String location, URI uri) {
        if (uri.isOpaque() || uri.getUserInfo() != null || uri.getRawQuery() != null || uri.getRawFragment() != null) {
            throw new SchemaException("invalid file schema location: " + location);
        }
        String host = uri.getHost();
        if (host != null && !host.isEmpty() && !"localhost".equalsIgnoreCase(host)) {
            throw new SchemaException("file URI has a non-local host: " + location);
        }
        String rawPath = uri.getRawPath();
        if (rawPath == null) {
            throw new SchemaException("invalid file schema location: " + location);
        }
        String lowerRawPath = rawPath.toLowerCase(Locale.ROOT);
        if (lowerRawPath.contains("%2f") || lowerRawPath.contains("%5c")) {
            throw new SchemaException("file URI contains an encoded path separator: " + location);
        }
        String path = uri.getPath();
        if (path == null || path.isEmpty() || path.indexOf('\0') >= 0) {
            throw new SchemaException("file URI does not contain a safe path: " + location);
        }
        if (path.length() >= 3 && path.charAt(0) == '/' && Character.isLetter(path.charAt(1)) && path.charAt(2) == ':') {
            path = path.substring(1);
        }
        try {
            Path resolvedPath = Path.of(path);
            if (!resolvedPath.isAbsolute()) {
                throw new SchemaException("file URI path is not absolute: " + location);
            }
            return resolvedPath.normalize();
        } catch (InvalidPathException e) {
            throw new SchemaException("invalid file schema location: " + location, e);
        }
    }

    private static String formatParseErrors(TomlParseResult document) {
        return document.errors().stream().map(Object::toString)
                .reduce((left, right) -> left + "; " + right).orElse("");
    }
}
